posthog/hogql/transforms/lazy_tables.py

```diff
@@ -0,0 +1,108 @@
+import dataclasses
+from typing import Callable, Dict, List, Optional, Set, Union
+
+from posthog.hogql import ast
+from posthog.hogql.resolver import resolve_symbols
+from posthog.hogql.visitor import TraversingVisitor
+
+
+def resolve_lazy_tables(node: ast.Expr, stack: Optional[List[ast.SelectQuery]] = None):
+    if stack:
+        # TODO: remove this kludge for old props
+        LazyTableResolver(stack=stack).visit(stack[-1])
+    LazyTableResolver(stack=stack).visit(node)
+
+
+class LazyTableResolver(TraversingVisitor):
+    def __init__(self, stack: Optional[List[ast.SelectQuery]] = None):
+        super().__init__()
+        self.stack_of_fields: List[List[ast.FieldSymbol]] = [[]] if stack else []
+
+    def _get_long_table_name(
+        self, select: ast.SelectQuerySymbol, symbol: Union[ast.TableSymbol, ast.LazyTableSymbol, ast.TableAliasSymbol]
+    ) -> str:
+        if isinstance(symbol, ast.TableSymbol):
+            return select.key_for_table(symbol)
+        elif isinstance(symbol, ast.TableAliasSymbol):
+            return symbol.name
+        elif isinstance(symbol, ast.LazyTableSymbol):
+            return f"{self._get_long_table_name(select, symbol.table)}__{symbol.field}"
+        else:
+            raise ValueError("Should not be reachable")
+
+    def visit_field_symbol(self, node: ast.FieldSymbol):
+        if isinstance(node.table, ast.LazyTableSymbol):
+            if len(self.stack_of_fields) == 0:
+                raise ValueError("Can't access a lazy field when not in a SelectQuery context")
+            self.stack_of_fields[-1].append(node)
+
+    def visit_select_query(self, node: ast.SelectQuery):
+        select_symbol = node.symbol
+        if not select_symbol:
+            raise ValueError("Select query must have a symbol")
+
+        # Collects each `ast.Field` with `ast.LazyTableSymbol`
+        field_collector: List[ast.FieldSymbol] = []
+        self.stack_of_fields.append(field_collector)
+
+        super().visit_select_query(node)
+
+        @dataclasses.dataclass
+        class JoinToAdd:
+            fields_accessed: Set[str]
+            join_function: Callable[[str, str, List[str]], ast.JoinExpr]
+            from_table: str
+            from_field: str
+            to_table: str
+
+        joins_to_add: Dict[str, JoinToAdd] = {}
+
+        for field in field_collector:
+            lazy_table = field.table
+            # traverse the lazy tables to a real table, then loop over them in reverse order to create the joins
+            joins_for_field: List = []
+            while isinstance(lazy_table, ast.LazyTableSymbol):
+                joins_for_field.append(lazy_table)
+                lazy_table = lazy_table.table
+            for lazy_table in reversed(joins_for_field):
+                from_table = self._get_long_table_name(select_symbol, lazy_table.table)
+                to_table = self._get_long_table_name(select_symbol, lazy_table)
```

Comment:
nit: I guess the context explains it, but it's slightly confusing here to see `lazy_table.table` and `lazy_table`